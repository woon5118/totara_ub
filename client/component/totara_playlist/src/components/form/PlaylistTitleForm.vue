<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Kian Nguyen <kian.nguyen@totaralearning.com>
  @module totara_playlist
-->
<template>
  <Form
    class="tui-playlistTitleForm"
    :vertical="true"
    input-width="full"
    @submit.prevent="submit"
  >
    <FormRow
      v-slot="{ id, label }"
      :label="$str('playlisttitle', 'totara_playlist')"
      :hidden="true"
    >
      <InputText
        :id="id"
        v-model="innerTitle"
        :placeholder="label"
        :aria-label="label"
        :maxlength="75"
        :autofocus="focusInput"
        name="playlist title"
        @keydown.native.esc="$emit('cancel', $event)"
      />
    </FormRow>

    <DoneCancelGroup
      :loading="submitting"
      :disabled="0 >= innerTitle.length"
      @done.prevent="submit"
      @cancel.prevent="$emit('cancel', $event)"
    />
  </Form>
</template>

<script>
import Form from 'tui/components/form/Form';
import FormRow from 'tui/components/form/FormRow';
import InputText from 'tui/components/form/InputText';
import DoneCancelGroup from 'totara_engage/components/buttons/DoneCancelGroup';

export default {
  components: {
    DoneCancelGroup,
    Form,
    FormRow,
    InputText,
  },

  props: {
    submitting: Boolean,
    title: {
      type: String,
      default: '',
    },
    /**
     * A flag to tell whether this form should focus on the input or not.
     */
    focusInput: Boolean,
  },

  data() {
    return {
      innerTitle: this.title,
    };
  },

  watch: {
    /**
     *
     * @param {String} value
     */
    title(value) {
      this.innerTitle = value;
    },
  },

  methods: {
    submit() {
      this.$emit('submit', { title: this.innerTitle });
    },
  },
};
</script>

<lang-strings>
  {
    "totara_playlist": [
      "playlisttitle"
    ]
  }
</lang-strings>

<style lang="scss">
.tui-playlistTitleForm {
  @include tui-font-body();
  display: flex;
  flex-direction: column;
  width: 100%;

  // Override form row
  .tui-formRow {
    &__desc {
      display: none;
    }

    &__action {
      margin: 0;
    }
  }
}
</style>
