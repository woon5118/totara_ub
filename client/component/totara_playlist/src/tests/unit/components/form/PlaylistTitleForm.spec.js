/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Kian Nguyen <kian.nguyen@totaralearning.com>
 * @module totara_playlist
 */

import PlaylistTitleForm from 'totara_playlist/components/form/PlaylistTitleForm';
import InputText from 'tui/components/form/InputText';
import { mount } from '@vue/test-utils';

describe('PlaylistTitleForm', () => {
  let form = null;

  beforeAll(() => {
    form = mount(PlaylistTitleForm, {
      attachToDocument: true,
      mocks: {
        $str(a, b) {
          return `${a}-${b}`;
        },
      },
      propsData: {
        title: 'Playlist Form',
        focusInput: true,
      },
    });
  });

  it('matches snapshot', () => {
    expect(form.element).toMatchSnapshot();
  });

  it('checks the aria label of the input field', () => {
    let input = form.find(InputText);
    expect(input.exists()).toBeTrue();
    expect(input.attributes('aria-label')).toBe(
      'playlisttitle-totara_playlist'
    );
  });

  it('should has the input focus', () => {
    let input = form.find(InputText);
    expect(input.exists()).toBeTrue();
    expect(input.element).toBe(document.activeElement);
  });

  it('checks the escape key event', () => {
    let input = form.find(InputText);
    expect(form.emitted().cancel).toBeUndefined();

    input.trigger('keydown.esc');
    expect(form.emitted().cancel).not.toBeUndefined();
  });
});
