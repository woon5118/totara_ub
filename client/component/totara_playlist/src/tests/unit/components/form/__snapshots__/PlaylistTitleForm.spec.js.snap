// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistTitleForm matches snapshot 1`] = `
<form
  autocomplete="off"
  class="tui-form tui-playlistTitleForm tui-form--vertical tui-form--inputWidth-full"
>
  <div
    class="tui-formRow tui-formRow--emptyDesc"
  >
    <div
      class="tui-formRow__inner"
    >
      <div
        class="tui-formRow__desc"
      >
        <label
          class="tui-formLabel tui-formLabel--hidden tui-formLabel--inline"
          for="id"
          id="id-label"
        >
          <!---->
          

  playlisttitle-totara_playlist
  
          <!---->
        </label>
        <!---->
      </div>
       
      <div
        class="tui-formRow__action tui-formRow__action--isStacked"
      >
        <input
          aria-label="playlisttitle-totara_playlist"
          class="tui-formInput"
          id="id"
          maxlength="75"
          name="playlist title"
          placeholder="playlisttitle-totara_playlist"
          type="text"
        />
      </div>
    </div>
  </div>
   
  <div
    class="tui-formBtnGroup tui-doneCancelGroup"
  >
    <button
      aria-label="done-totara_engage"
      class="tui-formBtn tui-formBtn--prim tui-formBtn--small"
      type="button"
    >
      
  done-totara_engage
  
      <!---->
       
      <!---->
    </button>
     
    <button
      class="tui-formBtn tui-formBtn--small"
      type="button"
    >
      
  cancel-core
  
      <!---->
       
      <!---->
    </button>
  </div>
</form>
`;
